slsim/Sources/QuasarCatalog/number_density_quasar.py

```diff
@@ -0,0 +1,246 @@
+    :param xi: (2) Best fit value of the observed evolution of the quasar luminosity function from SDSS DR3 survery (Richards et al. 2006)
+    :param z_star: (3) Best fit value of the observed evolution of the quasar luminosity function from SDSS DR3 survery (Richards et al. 2006)
+    :Note: You can adjust these values if needed
+
+    Returns:
+    :rtype: float
+    :return: M_star value.
+    """
+
+    denominator = (np.sqrt(np.exp(xi * z_value)) + np.sqrt(np.exp(xi * z_star))) ** 2
+    if denominator == 0:
+        return np.nan  # or any other suitable value to indicate an error
+    else:
+        return (
+            -20.90
+            + (5 * np.log10(h))
+            - (
+                2.5
+                * np.log10(
+                    np.exp(zeta * z_value) * (1 + np.exp(xi * z_star)) / denominator
+                )
+            )
+        )
+
+
+def dPhi_dM(M, z_value, alpha=-3.31, beta=-1.45, phi_star=5.34e-6 * (0.72**3)):
+    """Calculates dPhi_dM for a given M and redshift according to Eq (10) in Oguri &
+    Marshall (2010).
+
+    Parameters:
+    :param M: Absolute i-band magnitude.
+    :type M: float
+    :param alpha: Bright end slope of quasar luminosity density profile obtained from SDSS DR3 and 2dF surveys (Richards et al. 2005)
+    :param beta: Faint end slope of quasar luminosity density profile obtained from SDSS DR3 and 2dF surveys (Richards et al. 2005)
+    :param phi_star: Function to calculate the renormalization of the quasar luminosity function for a given h.
+
+    Returns:
+    :rtype: float
+    :return: dPhi_dM value.
+    """
+
+    if z_value > 3:
+        # Adjust the bright end slope for redshifts greater than 3 based on observations.
+        alpha_val = -2.58
+    else:
+        # Use the default value for the bright end slope if redshift is 3 or lower.
+        alpha_val = alpha
+
+    M_star_value = M_star(z_value)
+
+    denominator_dphi_dm = (10 ** (0.4 * (alpha_val + 1) * (M - M_star_value))) + (
+        10 ** (0.4 * (beta + 1) * (M - M_star_value))
+    )
+    term1 = phi_star / denominator_dphi_dm
+
+    return term1
+
+
+def cdf_fits_for_redshifts(M_values, random_redshift_values):
+    """Creates Cumulative Distribution Functions (CDF) plots for each randomly generated
+    redshift.
+
+    Parameters:
+    :param M_values: Array of absolute i-band magnitudes.
+    :type M_values: numpy.ndarray
+
+    :param random_redshift_values: Array of randomly generated redshift values.
+    :type random_redshift_values: numpy.ndarray
+
+    Returns:
+    :rtype: list
+    :return: List of legend handles for each CDF plot.
+    """
+
+    # Create an empty list to collect artists for the legend
+    legend_handles_cdf = []
+
+    # Plot the CDF vs. M curve for each randomly generated redshift
+    for random_redshift in random_redshift_values:
+        dPhi_dM_random_values = [dPhi_dM(M, random_redshift) for M in M_values]
```

Comment:
I am not sure I fully understand. You select one specific redshift but then go through all M values. Is this the intent?

Reply:
This is the intent (for now at least). For each redshift value in 'random_redshift_values', the code will calculate a 'dPhi_dM' value for each magnitude value in 'M_value.' This is done because the CDF depends on the magnitude distribution at that redshift, rather than the other way around.  Once we have the 'dPhi_dM' values for a specific redshift, we sort the magnitudes and compute the cumulative sum of the 'dPhi_dM' values. This allows us to isolate for M and carry on. 

 I believe as we apply cosmology to the code (such as the comoving volume formula) and integrate it to determine the number of quasars within a certain redshift, this method will be changed from using 'dPhi_dM' values to something that resembles reality. This is especially true for generating the redshift value since the code just randomly generates them. 

But for all moving cases, we will go through all M values.

Reply:
what we at the end want is:
- a function to generate redshifts for quasars based on the co-moving space density (within cuts in magnitudes) along a cone (with have the routines for it given a co-moving space density in bins of redshift)
- and then draw a magnitude from the PDF/CDF at that given redshift
We can discuss this at the meeting